src/torchmetrics/utilities/distributed.py

```diff
@@ -146,3 +147,65 @@ def gather_all_tensors(result: Tensor, group: Optional[Any] = None) -> List[Tens
+    """A specialized distributed sampler for evaluation (test and validation).
+
+    It is derived from the PyTorch DistributedSampler, with one core difference: it doesn't add extra samples to make
+    the data evenly divisible across devices. This is important while evaluating, as adding extra samples will screw
+    the results towards those duplicated samples.
+
+    Normally not adding the extra samples would lead to processes becoming out of sync, but this is handled by the
+    custom syncronization in Torchmetrics. Thus this sampler does not in general secure that distributed operations
+    are working outside of Torchmetrics.
+
+    Arguments are the same as DistributedSampler, and this implementation only overrides the __init__ method.
+
+    Args:
+        dataset: Dataset used for sampling.
+        num_replicas (int, optional): Number of processes participating in distributed training. By default,
+            :attr:`world_size` is retrieved from the current distributed group.
+        rank (int, optional): Rank of the current process within :attr:`num_replicas`. By default, :attr:`rank` is
+            retrieved from the current distributed group.
+        shuffle (bool, optional): If ``True`` (default), sampler will shuffle the indices.
+        seed (int, optional): random seed used to shuffle the sampler if :attr:`shuffle=True`. This number should be
+            identical across all processes in the distributed group.
+        drop_last (bool, optional): if ``True``, then the sampler will drop the tail of the data to make it evenly
+            divisible across the number of replicas.
+
+    For a full example on how to use this sampler, using both bare Pytorch but also Psytorch Lightning,
+    checkout the `distributed_evaluation.py` file in the examples folder.
+
+    Example::
+        The distributed sampler is always intended to be used in conjunction with a DataLoader:
+
+        >>> import torch
+        >>> from torch.utils.data import DataLoader, TensorDataset
+        >>> from torchmetrics.utilities.distributed import EvaluationDistributedSampler
+        >>> dataset = TensorDataset(torch.arange(10))
+        >>> dataloader = DataLoader(
+        ...   dataset, sampler=EvaluationDistributedSampler(dataset, num_replicas=2)
+        ... )  # doctest: +SKIP
+
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        num_replicas: Optional[int] = None,
+        rank: Optional[int] = None,
+        shuffle: bool = True,
+        seed: int = 0,
+        drop_last: bool = False,
+    ) -> None:
+        # From:
+        # https://github.com/pytorch/pytorch/issues/25162#issuecomment-1227647626
+        super().__init__(dataset=dataset, num_replicas=num_replicas, rank=rank, shuffle=shuffle, seed=seed)
+
+        len_dataset = len(self.dataset)  # type: ignore[arg-type]
+        if not self.drop_last and len_dataset % self.num_replicas != 0:
```

Comment:
the issue with this that it wouldn't necessarily work with validation, since not all ranks would reach the same distributed function calls and therefore time out which would kill the entire process. Also this would never work with FSDP, since some ranks have a batch more and for fsdp, not all processes would reach the forward syncing points also resulting in timeouts.

Reply:
I agree that in the context of Lightning this wouldn't work well, as it does not support Join (https://github.com/Lightning-AI/lightning/issues/3325)
FSDP also doesn't support join afaik (https://github.com/pytorch/pytorch/issues/64683)

But outside Lightning, and taking FSDP out of the equation, I agree this can work and is a good utility to have IMO. It also suits the metric design well, since synchronization is only necessary when all processes have finished collecting their statistics and `.compute()` can be called.